shell/platform/android/io/flutter/embedding/engine/systemchannels/TextInputChannel.java

```diff
@@ -0,0 +1,421 @@
+package io.flutter.embedding.engine.systemchannels;
+
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+import android.view.inputmethod.EditorInfo;
+
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import java.util.Arrays;
+import java.util.HashMap;
+
+import io.flutter.embedding.engine.dart.DartExecutor;
+import io.flutter.plugin.common.JSONMethodCodec;
+import io.flutter.plugin.common.MethodCall;
+import io.flutter.plugin.common.MethodChannel;
+
+/**
+ * {@link TextInputChannel} is a platform channel between Android and Flutter that is used to
+ * communicate information about the user's text input.
+ * <p>
+ * When the user presses an action button like "done" or "next", that action is sent from Android
+ * to Flutter through this {@link TextInputChannel}.
+ * <p>
+ * When an input system in the Flutter app wants to show the keyboard, or hide it, or configure
+ * editing state, etc. a message is sent from Flutter to Android through this {@link TextInputChannel}.
+ * <p>
+ * {@link TextInputChannel} comes with a default {@link io.flutter.plugin.common.MethodChannel.MethodCallHandler}
+ * that parses incoming messages from Flutter. Register a {@link TextInputMethodHandler} to respond
+ * to standard Flutter text input messages.
+ * <p>
+ * To handle all incoming text input messages in a custom manner, developers may override the
+ * default {@code MethodCallHandler} by using {@link #overrideDefaultMethodHandler(MethodChannel.MethodCallHandler)}.
```

Comment:
What is the intended use case for overriding the raw MethodCallHandler and its parsing logic?

Reply:
I offered it as an escape hatch.  I'll admit that I don't have a number of compelling use-cases, but I was just worried that a team might need to add some communication to a channel and they might otherwise have no reasonable way to do it.  Obviously the general approach should be to introduce new channels, not hijack existing ones, but I was just concerned about restricting these communication systems.  What do you think?

Reply:
My preference would be to remove this from the public API.  I don't think applications should add additional messages to the engine's internal channels.

Reply:
Done